node/src/peer_tracker.rs

```diff
@@ -1,308 +1,422 @@
+        }
+    }
+
+    pub(crate) fn is_full(&self) -> bool {
+        matches!(self, NodeKind::Full | NodeKind::Bridge)
+    }
+}
+
+impl Peer {
+    fn new(id: PeerId) -> Self {
+        Peer {
+            id,
+            connections: HashMap::new(),
+            protected: HashSet::new(),
+            trusted: false,
+            archival: false,
+            node_kind: NodeKind::Unknown,
+            // We start as disconnected
+            disconnected_at: Some(Instant::now()),
+        }
+    }
+
+    pub(crate) fn id(&self) -> &PeerId {
+        &self.id
+    }
+
+    pub(crate) fn is_connected(&self) -> bool {
+        !self.connections.is_empty()
+    }
+
+    pub(crate) fn is_trusted(&self) -> bool {
+        self.trusted
+    }
+
+    pub(crate) fn is_protected(&self) -> bool {
+        !self.protected.is_empty()
+    }
+
+    pub(crate) fn is_protected_for(&self, tag: u32) -> bool {
+        self.protected.contains(&tag)
+    }
+
+    pub(crate) fn is_archival(&self) -> bool {
+        self.archival
+    }
+
+    pub(crate) fn is_full(&self) -> bool {
+        self.node_kind.is_full()
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn node_kind(&self) -> NodeKind {
+        self.node_kind
+    }
+
+    pub(crate) fn best_ping(&self) -> Option<Duration> {
+        self.connections
+            .iter()
+            .flat_map(|(_, conn_info)| conn_info.ping)
+            .min()
     }
 }
 
 impl PeerTracker {
     /// Constructs an empty PeerTracker.
-    pub fn new(event_pub: EventPublisher) -> Self {
+    pub(crate) fn new(event_pub: EventPublisher) -> Self {
         PeerTracker {
-            peers: DashMap::new(),
+            peers: HashMap::new(),
+            protect_counter: HashMap::new(),
             info_tx: watch::channel(PeerTrackerInfo::default()).0,
             event_pub,
         }
     }
 
     /// Returns the current [`PeerTrackerInfo`].
-    pub fn info(&self) -> PeerTrackerInfo {
+    pub(crate) fn info(&self) -> PeerTrackerInfo {
         self.info_tx.borrow().to_owned()
     }
 
     /// Returns a watcher for any [`PeerTrackerInfo`] changes.
-    pub fn info_watcher(&self) -> watch::Receiver<PeerTrackerInfo> {
+    pub(crate) fn info_watcher(&self) -> watch::Receiver<PeerTrackerInfo> {
         self.info_tx.subscribe()
     }
 
-    /// Sets peer as discovered if this is it's first appearance.
+    pub(crate) fn peer(&self, peer_id: &PeerId) -> Option<&Peer> {
+        self.peers.get(peer_id)
+    }
+
+    pub(crate) fn peers(&self) -> impl Iterator<Item = &Peer> {
+        self.peers.values()
+    }
+
+    pub(crate) fn is_connected(&self, peer_id: &PeerId) -> bool {
+        self.peer(peer_id).is_some_and(|p| p.is_connected())
+    }
+
+    pub(crate) fn is_protected(&self, peer_id: &PeerId) -> bool {
+        self.peer(peer_id).is_some_and(|p| p.is_protected())
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn is_protected_for(&self, peer_id: &PeerId, tag: u32) -> bool {
```

Comment:
"Protected for tag" sounds weird to me somehow, but I'm not totally sure. Shouldn't it be "protected with tag"?

Reply:
renamed to `is_protected_with_tag`